### PR TITLE
Remove reference assembly and use EditorBrowsable.Never and runtime exceptions

### PR DESCRIPTION
**Pull request by bgavrilMS**
If you look at this, please review commit by commit.

### CODE REVIEW

File: src/client/Microsoft.Identity.Client/ApiConfig/AbstractConfidentialClientAcquireTokenParameterBuilder.cs [hunk 1]
```diff
@@ -12,19 +13,21 @@
 
 namespace Microsoft.Identity.Client
 {
-#if !ANDROID_BUILDTIME && !iOS_BUILDTIME && !WINDOWS_APP_BUILDTIME && !MAC_BUILDTIME // Hide confidential client on mobile platforms
-
     /// <summary>
     /// Base class for confidential client application token request builders
     /// </summary>
     /// <typeparam name="T"></typeparam>
+#if !SUPPORTS_CONFIDENTIAL_CLIENT
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/ApiConfig/AbstractConfidentialClientAcquireTokenParameterBuilder.cs:20]: I created a few new flags: 

- SUPPORTS_CONFIDENTIAL_CLIENT aka all platforms except mobile
- SUPPORTS_CUSTOM_CACHE - aka all platforms except mobile
- SUPPORTS_BROKER aka Android, iOS and UWP

File: src/client/Microsoft.Identity.Client/ApiConfig/AbstractConfidentialClientAcquireTokenParameterBuilder.cs [hunk 2]
```diff
@@ -12,19 +13,21 @@
...
+    [System.ComponentModel.EditorBrowsable(EditorBrowsableState.Never)]  // hide confidentail client on mobile
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/ApiConfig/AbstractConfidentialClientAcquireTokenParameterBuilder.cs:21]: This attribute ensures that Visual Studio (and potentially other IDEs) do not display this type in the list of suggestions. 

When applied to a method, it hides the method.
When applied to a class, it hides all instance methods on that class (not the static ones)

File: src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenInteractiveParameterBuilder.cs [hunk 3]
```diff
@@ -68,13 +68,10 @@ internal AcquireTokenInteractiveParameterBuilder WithCurrentSynchronizationConte
 
         internal AcquireTokenInteractiveParameterBuilder WithParentActivityOrWindowFunc(Func<object> parentActivityOrWindowFunc)
         {
-#if RUNTIME || NETSTANDARD_BUILDTIME
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenInteractiveParameterBuilder.cs:71]: No point in hiding internal stuff at buildtime.

File: src/client/Microsoft.Identity.Client/Microsoft.Identity.Client.csproj [hunk 4]
```diff
@@ -1,30 +1,258 @@
-﻿<Project Sdk="MSBuild.Sdk.Extras">
+﻿<Project  Sdk="MSBuild.Sdk.Extras">
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/Microsoft.Identity.Client.csproj:1]: No need to review this except the new flags SUPPORTS_CUSTOM_CACHE etc. 
It was just copied from directory.build.props

**Review by jennyf19 — APPROVED** [review]

**Review by trwalke — APPROVED** [review]
LGTM. I think these new directives are better. thanks

**Pull request merged**